Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+                else
+                {
+                    Start = start;
+                    End = end;
+
+                    if (compare == 0)
+                    {
+                        if (!(options.HasFlag(RangeOptions.IncludingStart)
+                              || options.HasFlag(RangeOptions.IncludingEnd)))
+                        {
+                            // empty
+                            Start = default(TValue);
+                            End = default(TValue);
+                            _options = RangeOptions.IsEmpty;
+                        }
+                        else
+                        {
+                            // ensure that everyting included
+                            _options |= RangeOptions.IncludingStart | RangeOptions.IncludingEnd;
+                        }
+                    }
+                }
+            }
+            else
+            {
+                Start = start;
+                End = end;
+            }
+        }
+
+        private Range(RangeValue start, RangeValue end)
+            : this(start.Value, end.Value,
+                (start.HasValue ? RangeOptions.HasStart : RangeOptions.None)
+                | (end.HasValue ? RangeOptions.HasEnd : RangeOptions.None)
+                | (start.Included ? RangeOptions.IncludingStart : RangeOptions.None)
+                | (end.Included ? RangeOptions.IncludingEnd : RangeOptions.None)
+                )
+        {
+        }
+
+        public bool HasStart
+        {
+            get { return _options.HasFlag(RangeOptions.HasStart); }
+        }
+
+        public bool HasEnd
+        {
+            get { return _options.HasFlag(RangeOptions.HasEnd); }
+        }
+
+        public bool IncludeStart
+        {
+            get { return _options.HasFlag(RangeOptions.IncludingStart); }
+        }
+
+        public bool IncludeEnd
+        {
+            get { return _options.HasFlag(RangeOptions.IncludingEnd); }
+        }
+
+        public bool IsEmpty
+        {
+            get { return _options.HasFlag(RangeOptions.IsEmpty); }
+        }
+
+        public bool IsFull
+        {
+            get { return _options == RangeOptions.None; }
+        }
+
+        private RangeValue StartValue
+        {
+            get { return new RangeValue(Start, HasStart, IncludeStart); }
+        }
+
+        private RangeValue EndValue
+        {
+            get { return new RangeValue(End, HasEnd, IncludeEnd); }
+        }
+
+        public int CompareTo(Range<TValue> other)
+        {
+            var current = this;
+            if (!current.HasStart)
+				if (!other.HasStart)
+					return 0;
+				else
+					return -1;
+
+			if (!other.HasStart)
+                return 1;
+
+            var compare = current.Start.CompareTo(other.Start);
+            if (compare == 0)
+            {
+                if (current.HasEnd && other.HasEnd)
+                {
+                    compare = current.End.CompareTo(other.End);
```

Comment:
So, here you include ends in comparison. However, you have ignored ends for left-unbounded regions. This logic looks inconsistent.